src/cqparts/display/cqparts_server.py

```diff
@@ -47,31 +49,59 @@ def display_callback(self, component):
-        if not isinstance(component, Component):
-            raise TypeError("given component must be a %r, not a %r" % (
-                Component, type(component)
-            ))
+        # get the server from the environment
+        server_url = os.environ[ENVVAR_SERVER]
+        # check that the server is running
+        try:
+            requests.get(server_url + '/status')
 
-        cp_name = type(component).__name__
+            # Verify Parameter(s)
+            # check that it is a component
+            from .. import Component
+            if not isinstance(component, Component):
+                raise TypeError("given component must be a %r, not a %r" % (
+                    Component, type(component)
+                ))
 
-        # create temporary folder
-        temp_dir = self._mkdir(tempfile.gettempdir(), 'cqpss')
-        temp_dir = self._mkdir(tempfile.gettempdir(), 'cqpss', cp_name)
+            # get the name of the object
+            cp_name = type(component).__name__
 
-        # export the files to the name folder
-        exporter = component.exporter('gltf')
-        exporter(
-            filename=os.path.join(temp_dir, 'out.gltf'),
-            embed=False,
-        )
+            # create temporary folder
+            #tmp_dir = self._mkdir(tempfile.gettempdir(), 'cqpss')
+            temp_dir = tempfile.mkdtemp()
+            base_dir = self._mkdir(temp_dir, cp_name)
 
-        # get the server from the environment
-        server_url = os.environ[ENVVAR_SERVER]
+            # export the files to the name folder
+            start_time = time.time()
+            exporter = component.exporter('gltf')
+            exporter(
+                filename=os.path.join(base_dir, 'out.gltf'),
+                embed=False,
+            )
+            finish_time = time.time()
+            duration = finish_time - start_time
+
+            # create the list of files to upload
+            file_list = os.listdir(base_dir)
+            file_load_list = []
+            for i in file_list:
+                # path of file to upload
+                file_name = os.path.join(base_dir, i)
+                # short reference to file
+                file_ref = os.path.join(cp_name, i)
+                # make dict for file upload
+                file_load_list.append(('objs',(file_ref,open(file_name, 'rb'))))
 
-        # notify the cq parts server
-        resp = requests.post(server_url + '/notify', data={
-            'name': cp_name,
-        })
+            # upload the files as multipart upload
+            requests.post(server_url + '/upload', files=file_load_list)
+            # notify the cq parts server
+            # TODO more data in post, bounding box , other data
+            requests.post(server_url + '/notify', data={
+                'duration': duration,
+                'name': cp_name,
+            })
+            # finally check that it's sane and delete
+            if os.path.exists(base_dir):
```

Comment:
Folder won't be deleted if an exception is caught.
I'd go with a pattern like:

```python
temp_dir = tempfile.mkdtemp()
try:
    pass  # do some stuff
except SomeError as e:  # optional
    log.error('something bad has happened')
    raise
finally:
    if os.path.isdir(temp_dir):
        shutil.rmtree(temp_dir)
```

Reply:
Agreed.